File: scripts/verify-linked-extension-independent.py
"""Independent native/object checks for the frozen linked/stack extension."""
import contextlib
import hashlib
import io
import json
from datetime import datetime, timezone
from pathlib import Path

directory = Path('scratch/linked-extension-independent')
payload = json.loads((directory / 'payload.json').read_text(encoding='utf-8'))
namespaces, outputs = {}, []
for name, example in payload['examples'].items():
    namespace = {'__name__': '__main__'}
    output = io.StringIO()
    with contextlib.redirect_stdout(output):
        exec(compile(example['code'], example['filename'], 'exec'), namespace)
    assert output.getvalue().strip() == example['output'].strip()
    namespaces[name] = namespace
    outputs.append({'name': name, 'stdout': output.getvalue().strip(), 'codeSha256': hashlib.sha256(example['code'].encode()).hexdigest()})

cycle = namespaces['cycle']
for case in payload['graphs']:
    nodes = [cycle['Node'](f'n{i}', 7) for i in range(len(case['next']))]
    for index, target in enumerate(case['next']):
        nodes[index].next = None if target is None else nodes[target]
    before = tuple(node.next for node in nodes)
    head = None if case['head'] is None else nodes[case['head']]
    entry = cycle['cycle_entry'](head)
    expected = None if case['entry'] is None else nodes[case['entry']]
    assert entry is expected
    assert cycle['entry_by_seen'](head) is expected
    assert cycle['cycle_size'](entry) == case['cycleLength']
    assert all(node.next is old for node, old in zip(nodes, before))

middle = namespaces['middle']
long_lengths = [0, 1, 2, 3, 6, 7, 41, 42, 1001, 1002, 10000]
for length in long_lengths:
    head, nodes = middle['make_chain']([7] * length)
    first, second = middle['middle'](head, True), middle['middle'](head)
    assert first is (nodes[(length - 1) // 2] if length else None)
    assert second is (nodes[length // 2] if length else None)
    before = tuple(node.next for node in nodes)
    left, right = middle['split_left_heavy'](head)
    collected = []
    for start in (left, right):
        current, part = start, []
        while current is not None:
            assert current not in part
            part.append(current)
            current = current.next
        collected.append(part)
    assert collected[0] + collected[1] == nodes
    assert len(collected[0]) == (length + 1) // 2
    assert len(collected[1]) == length // 2
    if length > 1:
        assert [i for i, (node, previous) in enumerate(zip(nodes, before)) if node.next is not previous] == [(length - 1) // 2]
    else:
        assert all(node.next is previous for node, previous in zip(nodes, before))

for case in payload['arrays']:
    assert namespaces['greater']['next_distances'](case['values'])[0] == case['strict']
    assert namespaces['greater']['next_distances'](case['values'], True)[0] == case['inclusive']
    left, right = namespaces['histogram']['smaller_boundaries'](case['heights'])
    assert left == case['left'] and right == case['right']
    area, witness = namespaces['histogram']['largest_rectangle'](case['heights'])
    assert area == case['area']
    if area:
        start, end, height = witness
        assert (end - start) * height == area
        assert height <= min(case['heights'][start:end])
    else:
        assert witness is None

changed = {
    'strict': namespaces['greater']['next_distances']([-5, -5, 0, -1, 0, 1])[0],
    'inclusive': namespaces['greater']['next_distances']([-5, -5, 0, -1, 0, 1], True)[0],
    'histogram': namespaces['histogram']['largest_rectangle']([0, 3, 3, 1, 4, 4, 0]),
    'bothBoundaries': namespaces['histogram']['smaller_boundaries']([0, 3, 3, 1, 4, 4, 0]),
}
assert changed['strict'] == [2, 1, 3, 1, 1, 0]
assert changed['inclusive'] == [1, 1, 2, 1, 1, 0]
assert changed['histogram'] == (8, (4, 6, 4))
for source in payload['sources']:
    assert hashlib.sha256(Path(source['path']).read_bytes()).hexdigest() == source['sha256']
result = {'checkedAt': datetime.now(timezone.utc).isoformat(), 'passed': True, 'sources': payload['sources'],
          'authorPacketSha256': payload['authorPacketSha256'], 'programs': outputs,
          'checks': {'allSuccessorGraphsAndHeadsThroughFiveNodes': len(payload['graphs']),
                     'middleModelStates': len(payload['middleCases']), 'nativeLongLengths': long_lengths,
                     'lengthSevenTernaryHistogramsAndSignedFutureArrays': len(payload['arrays']),
                     'frameMeaningAndStackPrefixInvariants': True, 'nodeIdentityAndNonmutation': True},
          'changedBrowserContracts': changed,
          'limits': 'Complementary exhaustive/object/witness checks support independently read proofs. No judge submissions or runtime benchmark.'}
(directory / 'native-results.json').write_text(json.dumps(result, indent=2) + '\n', encoding='utf-8')
print('Independent linked/stack native and frame/witness checks passed.')

File: scripts/verify-linked-extension-independent.mjs
import fs from 'node:fs';
import path from 'node:path';
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import { linkedCycleTrace, middleSplitState } from '../src/learn/data/linked-traversal-models.js';
import { nextGreaterTrace, histogramState } from '../src/learn/data/monotonic-stack-models.js';
import { linkedTraversalExamples } from '../src/learn/data/linked-traversal-examples.js';
import { monotonicStackExamples } from '../src/learn/data/monotonic-stack-examples.js';
const directory = 'scratch/linked-extension-independent';
fs.mkdirSync(directory, { recursive: true });
const hash = bytes => createHash('sha256').update(bytes).digest('hex');
const authorPath = 'docs/teaching/evidence/linked-traversal-extension-author-review.json';
const authorBytes = fs.readFileSync(authorPath), author = JSON.parse(authorBytes);
const sources = author.productionSources.map(source => {
  const bytes = fs.readFileSync(source.path); assert.equal(hash(bytes), source.sha256);
  const archive = path.join(directory, 'original-author-sources', source.path);
  fs.mkdirSync(path.dirname(archive), { recursive: true });
  if (!fs.existsSync(archive)) fs.writeFileSync(archive, bytes); else assert.equal(hash(fs.readFileSync(archive)), source.sha256);
  return { path: source.path, sha256: source.sha256, bytes: bytes.length };
});
const authorArchive = `${directory}/original-author-packet.json`;
if (!fs.existsSync(authorArchive)) fs.writeFileSync(authorArchive, authorBytes); else assert.equal(hash(fs.readFileSync(authorArchive)), hash(authorBytes));
function firstRepeat(next, head) {
  const visited = new Map(), walk = [];
  let current = head;
  while (current !== null && !visited.has(current)) {
    visited.set(current, walk.length); walk.push(current); current = next[current];
  }
  return { walk, entry: current, prefix: current === null ? null : visited.get(current),
    cycleLength: current === null ? 0 : walk.length - visited.get(current) };
}
const hop = (next, head, count) => {
  let current = head;
  for (let step = 0; step < count && current !== null; step += 1) current = next[current];
  return current;
};
const graphs = [];
for (let size = 0; size <= 5; size += 1) {
  for (let code = 0; code < (size + 1) ** size; code += 1) {
    const next = Array.from({ length: size }, (_, index) => { const digit = Math.floor(code / (size + 1) ** index) % (size + 1); return digit === size ? null : digit; });
    for (const head of [null, ...next.map((_, index) => index)]) {
      const saved = [...next], expected = firstRepeat(next, head), trace = linkedCycleTrace(next, head);
      assert.deepEqual(next, saved);
      assert.equal(trace.entry, expected.entry);
      assert.equal(trace.prefixLength, expected.prefix);
      assert.equal(trace.cycleLength, expected.cycleLength);
      for (const frame of trace.frames.filter(frame => frame.phase === 'detect')) {
        assert(frame.rounds > 0);
        assert.equal(frame.slow, hop(next, head, frame.rounds));
        assert.equal(frame.fast, hop(next, head, 2 * frame.rounds));
        assert.equal(frame.via, hop(next, head, 2 * frame.rounds - 1));
      }
      if (expected.entry !== null) {
        const firstMeeting = Math.max(expected.cycleLength, Math.ceil(expected.prefix / expected.cycleLength) * expected.cycleLength);
        assert.equal(trace.frames.find(frame => frame.phase === 'meeting').rounds, firstMeeting);
        assert.equal(trace.meeting, hop(next, head, firstMeeting));
        for (const frame of trace.frames.filter(frame => ['reset','entry-walk','entry'].includes(frame.phase))) {
          assert.equal(frame.slow, hop(next, head, frame.entrySteps));
          assert.equal(frame.fast, hop(next, trace.meeting, frame.entrySteps));
        }
      }
      graphs.push({ next, head, entry: trace.entry, cycleLength: trace.cycleLength, prefixLength: trace.prefixLength });
    }
  }
}
const middleCases = [];
for (let length = 0; length <= 12; length += 1) for (const policy of ['first','second']) for (const cut of [false,true]) {
  const state = middleSplitState(length, policy, cut);
  assert.equal(state.middle, length ? Math.floor((length - (policy === 'first' ? 1 : 0)) / 2) : null);
  const walk = head => firstRepeat(state.next, head).walk;
  if (cut) {
    const left = walk(state.leftHead), right = walk(state.rightHead);
    assert.deepEqual([...left,...right], Array.from({length},(_,i)=>i));
    assert.equal(left.length, Math.ceil(length / 2)); assert.equal(right.length, Math.floor(length / 2));
    assert.equal(left.filter(index => right.includes(index)).length, 0);
  }
  middleCases.push(state);
}
const arrays = [];
function directDistances(values, inclusive) {
  return values.map((value, index) => {
    const relative = values.slice(index + 1).findIndex(candidate => inclusive ? candidate >= value : candidate > value);
    return relative < 0 ? 0 : relative + 1;
  });
}
// All length-seven ternary histograms, plus signed versions, extend the author's shorter exhaustive set.
for (let serial = 0; serial < 3 ** 7; serial += 1) {
  const heights = Array.from({ length: 7 }, (_, index) => Math.floor(serial / 3 ** index) % 3);
  const histogram = histogramState(heights);
  const left = heights.map((height, index) => {
    for (let candidate = index - 1; candidate >= 0; candidate -= 1) if (heights[candidate] < height) return candidate;
    return -1;
  });
  const right = heights.map((height, index) => {
    for (let candidate = index + 1; candidate < heights.length; candidate += 1) if (heights[candidate] < height) return candidate;
    return heights.length;
  });
  let optimum = 0;
  for (let start = 0; start < heights.length; start += 1) for (let end = start + 1; end <= heights.length; end += 1) optimum = Math.max(optimum, (end - start) * Math.min(...heights.slice(start, end)));
  assert.deepEqual(histogram.leftTrace.boundaries, left); assert.deepEqual(histogram.rightTrace.boundaries, right);
  assert.equal(histogram.area, optimum);
  if (histogram.best) {
    assert.equal((histogram.best.end - histogram.best.start) * Math.min(...heights.slice(histogram.best.start, histogram.best.end)), optimum);
  } else assert.equal(optimum, 0);
  const values = heights.map(value => value - 1);
  const traces = [false,true].map(inclusive => nextGreaterTrace(values, inclusive));
  traces.forEach(trace => {
    assert.deepEqual(trace.distances, directDistances(values, trace.inclusive));
    assert.equal(trace.pushes, values.length); assert(trace.pops <= values.length);
    for (const frame of trace.frames.filter(frame => frame.phase === 'push')) {
      const prefix = values.slice(0, frame.current + 1), expected = prefix.flatMap((value, index) => directDistances(prefix, trace.inclusive)[index] === 0 ? [index] : []);
      assert.deepEqual(frame.stack, expected);
    }
  });
  arrays.push({ heights, left, right, area: optimum, values, strict: traces[0].distances, inclusive: traces[1].distances });
}
const payload = { checkedAt: new Date().toISOString(), sources, authorPacketSha256: hash(authorBytes),
  graphs, middleCases, arrays, examples: { ...linkedTraversalExamples, ...monotonicStackExamples } };
fs.writeFileSync(`${directory}/payload.json`, JSON.stringify(payload));
execFileSync('scratch/lesson-tools/Scripts/python.exe', ['scripts/verify-linked-extension-independent.py'], { stdio: 'inherit', timeout: 120000 });
console.log(JSON.stringify({ graphs: graphs.length, middleCases: middleCases.length, changedSevenElementArrays: arrays.length, sourceHashes: sources.length }));
